### PR TITLE
etcdv3 registry

### PR DESCRIPTION
**Pull request by HeavyHorst**
etcdv3 registry backend :)

the watcher is not fully working at the moment. I need to find a way to get the service on a delete event.
I will complete this tomorrow.

### CODE REVIEW

**Comment by asim** [comment]
What's the reason for having this as a separate plugin? Is this incompatible with the previous version of etcd?

**Comment by HeavyHorst** [comment]
the api has completely changed -> [(https://github.com/coreos/etcd/releases/tag/v3.0.0-beta.0)]

> Some highlights of the new etcd3 API:
> 
> Flat binary keyspace
> No key-value hierarchy / directories
> Support query/watch by prefix
> Transactions: combine multiple requests into one operation
> Leases: a single TTL for sets of keys
> Maintenance/Alarm: protect etcd against unexpected shutdown, with disk quota limits

they also switched from http to grpc and http/2.

**Comment by asim** [comment]
I'll be curious to know how they actually end up versioning this. My first thought goes to what the right location for a plugin like this is since we're talking about versions of the same thing. Whether it should be under registry/etcd/v3 or registry/etcdv3. 

Besides all that. Appreciate the contribution :)

**Comment by asim** [comment]
Can you add a test to this? Realised we need one for the zookeeper registry as well which should be easy considering the client library has a test cluster function.

**Comment by HeavyHorst** [comment]
sure, i will add the tests soon !

File: registry/etcdv3/watcher.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+					Service: service,
+				}, nil
+			}
+		}
+	}
+	return nil, errors.New("could not get next")
+}
+
+func (ew *etcdv3Watcher) Stop() {
+	ew.once.Do(func() {
```
Comment (asim) [inline registry/etcdv3/watcher.go]: Do this instead and get rid of sync.Once

```
select {
case <-ew.stop:
  return
default:
  close(ew.stop)
}
```

File: registry/etcdv3/watcher.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+		if wresp.Err() != nil {
+			return nil, wresp.Err()
+		}
+		for _, ev := range wresp.Events {
+			service := decode(ev.Kv.Value)
+			switch ev.Type {
+			case clientv3.EventTypePut, clientv3.EventTypeDelete:
+				var action string
+
+				if ev.Type == clientv3.EventTypePut {
```
Comment (asim) [inline registry/etcdv3/watcher.go]: Maybe use switch on ev.Type since its being looked at multiple times.

File: registry/etcdv3/etcdv3.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+	defer cancel()
+
+	resp, err := e.client.Grant(ctx, int64(options.TTL.Seconds()))
+	if err != nil {
+		return err
+	}
+
+	for _, node := range s.Nodes {
+		service.Nodes = []*registry.Node{node}
+		_, err := e.client.Put(ctx, nodePath(service.Name, node.Id), encode(service), clientv3.WithLease(resp.ID))
```
Comment (asim) [inline registry/etcdv3/etcdv3.go]: Don't ignore encode errors.

File: registry/etcdv3/etcdv3.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+	}
+
+	if len(rsp.Kvs) == 0 {
+		return nil, registry.ErrNotFound
+	}
+
+	serviceMap := map[string]*registry.Service{}
+
+	for _, n := range rsp.Kvs {
+		sn := decode(n.Value)
```
Comment (asim) [inline registry/etcdv3/etcdv3.go]: Don't ignore decode errors. If this is nil it will panic.

File: registry/etcdv3/etcdv3.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), e.options.Timeout)
+	defer cancel()
+
+	for _, node := range s.Nodes {
+		// cache the value for the watcher
+		resp, err := e.client.Get(ctx, nodePath(s.Name, node.Id))
+		if err != nil {
+			return err
+		}
+		for _, ev := range resp.Kvs {
```
Comment (asim) [inline registry/etcdv3/etcdv3.go]: Nitpick. Needs space between if statement and for statement.

File: registry/etcdv3/etcdv3.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+			lrsp, err := e.client.Grant(context.TODO(), 5)
+			if err != nil {
+				return err
+			}
+			_, err = e.client.Put(ctx, path.Join(cachePrefix, string(ev.Key)), string(ev.Value), clientv3.WithLease(lrsp.ID))
```
Comment (asim) [inline registry/etcdv3/etcdv3.go]: Space here too.

**Pull request merged**